// toxiproxy-node-client/dist/Proxy.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const Toxic_1 = require("./Toxic");
class Proxy {
    constructor(toxiproxy, body) {
        this.api = toxiproxy.api;
        this.toxiproxy = toxiproxy;
        const { name, listen, upstream, enabled } = body;
        this.name = name;
        this.listen = listen;
        this.upstream = upstream;
        this.enabled = enabled;
    }
    getToxiproxy() {
        return this.toxiproxy;
    }
    toJson() {
        return {
            enabled: this.enabled,
            listen: this.listen,
            name: this.name,
            upstream: this.upstream
        };
    }
    getHost() {
        return this.toxiproxy.host;
    }
    getPath() {
        return `${this.getHost()}/proxies/${this.name}`;
    }
    /**
     * Deletes the proxy from the server.
     */
    remove() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.api.delete(this.getPath());
        });
    }
    /**
     * Updates the existing proxy on the server (e.g. disable an enabled proxy).
     *
     * @returns updated proxy
     */
    update(body) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.api.post(this.getPath(), body)
                .then((response) => {
                return new Proxy(this.toxiproxy, response.data);
            });
        });
    }
    /**
     * Adds a new toxic to the proxy.
     *
     * @param body toxic attributes
     * @returns toxic
     */
    addToxic(body) {
        return __awaiter(this, void 0, void 0, function* () {
            const result = yield this.api.post(`${this.getPath()}/toxics`, body)
                .then((response) => {
                return response.data;
            });
            return new Toxic_1.default(this.api, this.getPath(), result);
        });
    }
    /**
     * Gets the toxic from the proxy.
     *
     * @param name toxic name
     * @returns toxic
     */
    getToxic(name) {
        return __awaiter(this, void 0, void 0, function* () {
            const result = yield this.api.get(`${this.getPath()}/toxics/${name}`)
                .then((response) => {
                return response.data;
            });
            return new Toxic_1.default(this.api, this.getPath(), result);
        });
    }
}
exports.default = Proxy;
