// toxiproxy-node-client/dist/TestHelper.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.removeAllProxies = exports.createToxic = exports.createProxy = exports.toxiproxyUrl = void 0;
const Toxiproxy_1 = require("./Toxiproxy");
exports.toxiproxyUrl = "http://localhost:8474";
const createProxy = (name) => __awaiter(void 0, void 0, void 0, function* () {
    const toxiproxy = new Toxiproxy_1.default(exports.toxiproxyUrl);
    const body = {
        listen: "localhost:0",
        name: name,
        upstream: "localhost:6379"
    };
    const proxy = yield toxiproxy.createProxy(body);
    return { proxy, toxiproxy };
});
exports.createProxy = createProxy;
const createToxic = (proxy, type, attributes) => __awaiter(void 0, void 0, void 0, function* () {
    const body = {
        attributes: attributes,
        type: type
    };
    return yield proxy.addToxic(body);
});
exports.createToxic = createToxic;
function removeAllProxies() {
    return __awaiter(this, void 0, void 0, function* () {
        const toxiproxy = new Toxiproxy_1.default(exports.toxiproxyUrl);
        const proxies = yield toxiproxy.getAll();
        const promises = [];
        for (const proxyName in proxies) {
            if (proxies.hasOwnProperty(proxyName)) {
                const proxy = proxies[proxyName];
                yield proxy.remove();
            }
        }
        return Promise.all(promises);
    });
}
exports.removeAllProxies = removeAllProxies;
