// toxiproxy-node-client/dist/Toxiproxy.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ToxiproxyError = void 0;
const axios_1 = require("axios");
const Proxy_1 = require("./Proxy");
class ToxiproxyError {
    constructor(name, message, stack) {
        this.name = name;
        this.message = message;
        this.stack = stack;
    }
}
exports.ToxiproxyError = ToxiproxyError;
class Toxiproxy {
    constructor(host) {
        this.api = axios_1.default.create();
        this.api.interceptors.response.use((response) => response, (error) => {
            if (error instanceof axios_1.AxiosError) {
                // Return a simplified error object to avoid "TypeError: Converting circular structure to JSON".
                // Really not sure why this is not the default behavior.
                return Promise.reject(new ToxiproxyError(error.name, error.message, error.stack));
            }
            return Promise.reject("Unknown error");
        });
        this.host = host;
    }
    getApi() {
        return this.api;
    }
    createProxy(body) {
        return __awaiter(this, void 0, void 0, function* () {
            const result = yield this.api.post(`${this.host}/proxies`, body);
            return new Proxy_1.default(this, result.data);
        });
    }
    populate(body) {
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield this.api.post(`${this.host}/populate`, body);
            const proxies = {};
            for (const proxyResponse of res.data.proxies) {
                proxies[proxyResponse.name] = new Proxy_1.default(this, proxyResponse);
            }
            return proxies;
        });
    }
    get(name) {
        return __awaiter(this, void 0, void 0, function* () {
            const result = yield this.api.get(`${this.host}/proxies/${name}`);
            return new Proxy_1.default(this, result.data);
        });
    }
    getVersion() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.api.get(`${this.host}/version`);
        });
    }
    reset() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.api.post(`${this.host}/reset`);
        });
    }
    getAll() {
        return __awaiter(this, void 0, void 0, function* () {
            const responses = yield this.api.get(`${this.host}/proxies`);
            const proxies = {};
            for (const name in responses.data) {
                proxies[name] = new Proxy_1.default(this, responses.data[name]);
            }
            return proxies;
        });
    }
}
exports.default = Toxiproxy;
