// toxiproxy-node-client/dist/Toxic.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
class Toxic {
    constructor(api, proxyPath, body) {
        this.api = api;
        this.proxyPath = proxyPath;
        this.name = body.name;
        this.type = body.type;
        this.stream = body.stream;
        this.toxicity = body.toxicity;
        this.attributes = body.attributes;
    }
    toJson() {
        return {
            attributes: this.attributes,
            name: this.name,
            stream: this.stream,
            toxicity: this.toxicity,
            type: this.type
        };
    }
    getPath() {
        return `${this.proxyPath}/toxics/${this.name}`;
    }
    /**
     * Deletes the toxic from the server.
     */
    remove() {
        return __awaiter(this, void 0, void 0, function* () {
            return this.api.delete(this.getPath());
        });
    }
    /**
     * Updates the toxic's attributes on the server.
     *
     * @param body Toxic attributes to update
     * @returns Toxic with updated attributes
     */
    update(body) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.api.post(this.getPath(), body)
                .then((response) => {
                return new Toxic(this.api, this.proxyPath, response.data);
            });
        });
    }
}
exports.default = Toxic;
